Have SpECTRE control system act more like SpECs control system

knelli2 opened the pull request:
## Proposed changes

Also write the horizon centers to disk every time the control system horizon finder is called because this is a helpful diagnostic.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Output centers after control system horizon find

knelli2 committed:
Don't write control signal in control system output

knelli2 commented on src/ControlSystem/ApparentHorizons/Measurements.hpp:
```diff
@@ -69,7 +72,8 @@ struct BothHorizons : tt::ConformsTo<protocols::Measurement> {
           intrp::callbacks::ErrorOnFailedApparentHorizon;
       using post_horizon_find_callbacks =
-          tmpl::list<control_system::RunCallbacks<FindHorizon, ControlSystems>>;
+          tmpl::list<control_system::RunCallbacks<FindHorizon, ControlSystems>,
+                     ::ah::callbacks::ObserveCenters<InterpolationTarget>>;
```
Hi.

wthrowe commented on src/ControlSystem/Tags.hpp:
```diff
@@ -96,6 +96,18 @@ struct WriteDataToDisk {
+/// Option tag that determines how many measurements will occur per control
+/// system update.
+struct MeasurementsPerUpdate {
+  using type = int;
```
Set a lower bound of 1?

wthrowe commented on src/ControlSystem/Trigger.hpp:
```diff
@@ -148,15 +161,53 @@ class Trigger : public DenseTrigger {
+    // Just take the average of the two things we are comparing. They'll always
+    // be super close
+    const double scale =
+        0.5 * (calculated_next_measurement + min_fot_expr_time);
```
Pass zero as the scale.  The scale is to correct for error caused by intermediate results canceling in the calculation.

wthrowe commented on src/ControlSystem/Trigger.hpp:
```diff
@@ -148,15 +161,53 @@ class Trigger : public DenseTrigger {
+
+    // If the expiration time is infinity, don't bother checking if it's equal
+    // to the min measure time. We already returned above if it is infinity
+    if (min_fot_expr_time == std::numeric_limits<double>::infinity()) {
```
Is the false branch tested?  Codecov thinks it isn't.

wthrowe commented on src/ControlSystem/Trigger.hpp:
```diff
@@ -148,15 +161,53 @@ class Trigger : public DenseTrigger {
+    double min_fot_expr_time = std::numeric_limits<double>::infinity();
+    for (const auto& [name, fot] : functions_of_time) {
+      (void)name;
+      min_fot_expr_time = std::min(min_fot_expr_time, fot->time_bounds()[1]);
```
Are you sure that the expiration times are in a consistent state on all elements when they run this function?  It looks like this can run any time after the measurement timescales update, which I don't think is sequenced with respect to the main FoTs.

knelli2 replied:
Good catch. I think my new implementation catches this, but let me know if you'd like me to change it. It isn't the prettiest solution.

wthrowe replied:
Thinking about this a bit more, I think this can fail in the other direction as well:  If some element that doesn't need to supply data is more than one measurement behind the elements actually used in a measurement, it could receive the FoT update before scheduling the trigger at the update time.

Is there some other way we could obtain the update time?  The expiration time was really not intended to be used as data.

knelli2 replied:
I believe these concerns are now moot because the functions of time aren't updated at the expiration time anymore.

wthrowe replied:
I don't think that's relevant?  The numbers change a bit, but I don't see how it fixes the race.

knelli2 replied:
Hmmm you're right. We have the `control_system::Tags::MeasurementsPerUpdate` in the cache. What if we store which measurement we are on in the trigger, and then when we are at the update time (which is `current_measurement == measurements per update`), we store the expiration time in the trigger and use that expiration time. This way there's no race condition because nothing has been updated yet because we have to actually do the measurement the functions of time can be updated. There'd have to be a special case for t=0, but that's doable. What do you think @wthrowe?

wthrowe replied:
Do we necessarily have to have done the measurement?  Someone has to have done the measurement, certainly, but a horizon-find might not wait for data from the wave zone.

What is the purpose of this roundoff check?  Can we just remove it?

knelli2 replied:
I don't think we can get rid of the roundoff check. We were having all these issues syncing up floating point numbers because the expiration time is calculated on the control system components while the next measurement time is calculated on the elements and those weren't lining up.

Given that, I just need an expiration time that's guaranteed to be valid and not subject to race conditions. This was the simplest way I could think of at the moment. If you have a better way, I'm open to suggestions, but I'd really like to get this in quickly so we can do BBH testing with it.

wthrowe replied:
But why do they have to line up?

knelli2 replied:
Originally they had to because we updated at the expiration time. But now since we update before the expiration time, I guess they don't have to line up exactly. If the measurement time is less than the expiration time by roundoff (or equal to exactly), nothing happens because everything is valid. If the measurement time is greater than the expiration time by roundoff, then it will wait for the functions of time to update which is now guaranteed to happen because we updated the previous measurement.

I've tested a BBH where I got rid of the roundoff check and it's run fine so far up to 30M which is already a lot of updates.

wthrowe commented on tests/Unit/Helpers/ControlSystem/SystemHelpers.hpp:
```diff
@@ -557,13 +566,28 @@ struct SystemHelper {
+    };
+    double dt = calculate_dt(time);
+
+    // using obs_writer = typename Metavars::observer_component;
```
Remove.

wthrowe commented on src/ControlSystem/Trigger.hpp:
```diff
@@ -143,6 +143,55 @@ class Trigger : public DenseTrigger {
+      // If it is equal to the current time, then at least one of the functions
+      // of time hasn't been updated yet and we must wait until they are all
+      // updated so we set a callback.
+      if (must_check_fot_validity_) {
```
Why is this check done conditionally?  It looks like `next_measurement` always needs it to be true, and it doesn't look terribly expensive.

knelli2 reviewed (commented):
After discussing with people on 10/18/22, it was decided that instead of updating the functions of time at the expiration time, they should be updated one measurement before the expiration time. The new fixups implement these changes.

wthrowe commented on src/ControlSystem/ExpirationTimes.hpp:
```diff
@@ -0,0 +1,122 @@
+ *
+ * To protect against bad inputs, if the initial expiration time that is
+ * calculated is smaller than the initial time step, then the expiration time is
+ * simply set to the initial time step. However, the MeasurementTimescales have
```
You may need to revert d8ac31aa3f57aa9ae9ba18262e3b17ebb16fe557 to avoid a potential problem with self-start evaluation times.  It allows self-start to take initialization steps at times after the end of the first real step.  If you do, I have no objection to that happening.

knelli2 replied:
I've been using this for all the runs I've been doing since this was originally merged over a year ago, and it hasn't been an issue so far so I'll leave it for now. If it becomes an issue, we can deal with it then.

wthrowe commented on src/ControlSystem/Trigger.hpp:
```diff
@@ -148,15 +161,49 @@ class Trigger : public DenseTrigger {
+
+    // If the expiration time is infinity, don't bother checking if it's equal
+    // to the min measure time. We already returned above if it is infinity
+    if (min_fot_expr_time == std::numeric_limits<double>::infinity()) {
```
Is there a case where having a finite measurement time makes sense when the FoTs never expire?

knelli2 replied:
I can't think of a regular use case. I could imagine some convoluted debugging scenario where we don't want to update the functions of time but still do the measurements? But that seems so specific that it should be done on a branch and not have the functionality on develop.

I'll remove this part.

knelli2 committed:
Revamp control system Trigger test

Also avoid FPE when measurement timescale is infinity

knelli2 committed:
Have SpECTRE control system act more like SpECs control system

The way it acts more like SpECs is that all control systems use
the same measurements and use a specific number of measurements
before it updates the functions of time.

It differs from SpECs in that the Nth derivative of the
functions of time aren't changed at the update time, but rather
one measurement later. This is done to avoid extra global syncs.

knelli2 commented:
@wthrowe All those fixups were getting kind of confusing so I squashed everything. Because we don't have to worry about roundoff for the expiration time in the Trigger anymore, there were basically no changes needed to the Trigger. One small thing about infinity but I factored that out into it's own commit.

I didn't change much else. The main things to look at are `ExpirationTimes.?pp`, `FunctionsOfTimeInitialize.hpp`, and `MeasurementTimescales.hpp` (and their corresponding tests). Basically everything else is the same as the last time you reviewed this PR.

wthrowe commented on src/ControlSystem/ExpirationTimes.hpp:
```diff
@@ -0,0 +1,140 @@
+template <size_t DerivOrder>
+DataVector calculate_measurement_timescales(
+    const ::Controller<DerivOrder>& controller, const ::TimescaleTuner& tuner,
+    const int measurements_per_update);
```
Sorry if this was here before and I missed it, but if you're going to do it this way you should explicitly instantiate `calculate_measurement_timescales` so that this file doesn't depend on the details of implementations of functions elsewhere causing implicit instantiations.  Better would be to break the include loop.

wthrowe reviewed (changes requested):
Also rebase for a conflict.

knelli2 commented:
@wthrowe Rebased to fix conflicts and just separated out the `calculate_measurement_timescales` function into a new file to avoid include loops.

nilsdeppe commented on src/ControlSystem/Trigger.hpp:
```diff
@@ -143,20 +143,27 @@ class Trigger : public DenseTrigger {
-               });
+    const double min_measure_time = tmpl::as_pack<ControlSystems>(
+        [&measurement_timescales, &time](auto... control_systems) {
+          // LCOV_EXCL_START
```
Do you understand why you need to exclude all this from test coverage? It seems like this should be covered since it's the core part of this function

knelli2 replied:
No not really....I know for a fact the code is tested (the trigger won't work without it), but for some reason codecov says it isn't tested?? Maybe has something to do with the `tmpl::as_pack`? I just added the `LCOV*` to silence codecov's false warning.

nilsdeppe replied:
Alright, sounds good :)

nilsdeppe commented on src/ControlSystem/CalculateMeasurementTimescales.hpp:
```diff
@@ -0,0 +1,36 @@
+ * that the largest the measurement timescale can be is \f$\tau_\mathrm{m} =
+ * \tau_\mathrm{update} / N\f$. To ensure that we have sufficient measurements,
+ * we calculate the measurement timescales as \f$\tau_\mathrm{m} =
+ * \tau_\mathrm{update} / (N+1)\f$
```
Could you add that the `N+1` factor is `measurements_per_update` and is set elsewhere? I kept looking for an `N+1` 🙃

nilsdeppe commented on src/ControlSystem/ExpirationTimes.hpp:
```diff
@@ -0,0 +1,135 @@
+ * \ingroup ControlSystemGroup
+ * \brief Calculate the next expiration time for the FunctionsOfTime.
+ *
+ * This is done by adding to the current time one old minimum measurement
```
Could you write a math equation here? The words are a bit, well, wordy 🙃

nilsdeppe commented on src/ControlSystem/ExpirationTimes.hpp:
```diff
@@ -0,0 +1,135 @@
+ * subsequent measurement (and by that time, most, if not all, functions of time
+ * should have been updated because an entire horizon find happened in the
+ * meantime). If the expiration time was earlier than the next measurement, we'd
+ * have to pause the Algorithm on the elements and wait until all the functions
```
maybe `elements` -> `DG elements`

nilsdeppe commented on src/ControlSystem/ExpirationTimes.hpp:
```diff
@@ -0,0 +1,135 @@
+ * function of time expiration time, our measurement timescales can't be valid
+ * and we must wait for updated ones. To achieve this, we set the measurement
+ * expiration time *before* the function of time expiration time, but *after*
+ * the previous measurement (the update measurement). The factor of one half is
```
Can you add an equation here? The 1/2 is in the code, but the comment doesn't really make that clear since the formula isn't explicitly given

nilsdeppe commented on src/ControlSystem/ExpirationTimes.hpp:
```diff
@@ -0,0 +1,135 @@
+ *
+ * To protect against bad inputs, if the initial expiration time that is
+ * calculated is smaller than the initial time step, then the expiration time is
+ * simply set to the initial time step. However, the MeasurementTimescales have
```
Given this caveat and semi-silent error if both are invalid, would it be better to error on bad inputs? Or set the initial measurement time to 1/2 the initial time step?

nilsdeppe commented on src/ControlSystem/UpdateControlSystem.hpp:
```diff
@@ -130,84 +118,96 @@ struct UpdateControlSystem {
-    // because we still want the averager to be up to date even if we aren't
-    // updating at this time
+    // Update the averager. We do this for every measurement because we still
+    // want the averager to be up to date even if we aren't updating at this
```
[optional] `up-to-date`?

nilsdeppe reviewed (changes requested):
Overall LGTM, just some requests for more equations/dox :) Please rebase and squash immediately!
